### ISSUE TITLE
Initial draft proposal for 'oneof' support in generated code

### ISSUE THREAD

**Issue by danrice-square** [square/wire#203]
Here's a quick stab at what the generated code for `oneof` could look like (some irrelevant methods omitted). Comments are welcome.

```proto
message OneOfMessage {

  oneof choice {
      int32 foo = 1;
      string bar = 2;
  }
}
```

```java
public final class OneOfMessage extends Message {

  @ProtoField(tag = 1, type = INT32)
  public final Integer foo;

  @ProtoField(tag = 2, type = STRING)
  public final String bar;

  public enum ChoiceCase implements ProtoEnum {
    CHOICE_NOT_SET(0),
    FOO(1),
    BAR(2);

    int value;

    ChoiceCase(int value) {
      this.value = value;
    }

    public int getValue() {
      return value;
    }

    static ChoiceCase valueOf(int value) {
      switch (value) {
        case 0: return CHOICE_NOT_SET;
        case 1: return FOO;
        case 2: return BAR;
      }
      return null;
    }
  };

  public final ChoiceCase choiceCase;

  public OneOfMessage(Integer foo, String bar, ChoiceCase choiceCase) {
    this.foo = foo;
    this.bar = bar;
    this.choiceCase = choiceCase;
  }

  private OneOfMessage(Builder builder) {
    this(builder.foo, builder.bar, builder.choiceCase);
    setBuilder(builder);
  }

  public static final class Builder extends Message.Builder<OneOfMessage> {

    public Integer foo;
    public String bar;
    public ChoiceCase choiceCase = ChoiceCase.CHOICE_NOT_SET;

    public Builder() {
    }

    public Builder(OneOfMessage message) {
      super(message);
      if (message == null) return;
      this.foo = message.foo;
      this.bar = message.bar;
      this.choiceCase = message.choiceCase;
    }

    public Builder foo(Integer foo) {
      this.bar = null;
      this.foo = foo;
      this.choiceCase = foo == null ? ChoiceCase.CHOICE_NOT_SET : ChoiceCase.FOO;
      return this;
    }

    public Builder bar(String bar) {
      this.foo = null;
      this.bar = bar;
      this.choiceCase = bar == null ? ChoiceCase.CHOICE_NOT_SET : ChoiceCase.BAR;
      return this;
    }

    @Override
    public OneOfMessage build() {
      return new OneOfMessage(this);
    }
  }
}
```

**Comment by JakeWharton** [comment]
Nice. Why the `Case` suffix? Is that what `protoc` does?

**Comment by danrice-square** [comment]
Yes. I'm definitely open to a better naming convention.

**Comment by danrice-square** [comment]
How about `ChoiceEnum`?

**Comment by JakeWharton** [comment]
Does it need suffixed at all? I haven't played with them enough but I assume the scoping rules are the same as a field so something like:
```proto
oneof choice {
  string choice = 1;
  string choice2 = 2;
}
```
would fail.

So why not just retain the raw name of the `oneof` like you would any other field?

**Comment by JakeWharton** [comment]
Fixed by #205
